### ISSUE TITLE
Problems with mixed precision between fields and particles

### ISSUE THREAD

**Issue by jwise77** [enzo-project/enzo-e#197]
When I was debugging the particles in the IsolatedGalaxy test, I discovered that it would not run when the user-specified particle attribute precision did not match the precision of `enzo_float`. These crashes occur when the particle attribute is cast into the `enzo_float` type in a call like
```c++
  px    = (enzo_float *) particle.attribute_array(it, ia_x, ib);
```
that would result in garbled data when `type(enzo_float) != type(particle.attribute_type)`.

In the IsolatedGalaxy initializer, I worked around this issue by writing the following routine
```c++
template <class T>
void EnzoInitialIsolatedGalaxy::assign_attribute_value
(Particle * particle, T value, int it, int ia, int gindex)
{
  if (ia < 0) { // undefined
    return;
  }

  int ib = 0;
  int ipp = 0;

  particle->index(gindex, &ib, &ipp);

  // Get pointer of the attribute array
  char * attr = particle->attribute_array(it,ia,ib);

  // Get type of attribute array
  int type = particle->attribute_type(it,ia);

  if (cello::type_is_float(type)) {
    if (type == type_single) {
      float * attr_f = (float*) attr;
      attr_f[ipp] = value;
    } else if (type == type_double) {
      double * attr_d = (double*) attr;
      attr_d[ipp] = value;
    } else if (type == type_quadruple) {
      long double * attr_q = (long double*) attr;
      attr_q[ipp] = value;
    } else {
      ERROR1("EnzoInitialIsolatedGalaxy.C // assign_attribute_value():",
      "Unknown particle attribute float type %d", type);
    }
  } else if (cello::type_is_int(type)) {
    if (type == type_int8) {
      int8_t * attr_8 = (int8_t *) attr;
      attr_8[ipp] = value;
    } else if (type == type_int16) {
      int16_t * attr_16 = (int16_t *) attr;
      attr_16[ipp] = value;
    } else if (type == type_int32) {
      int32_t * attr_32 = (int32_t *) attr;
      attr_32[ipp] = value;
    } else if (type == type_int64) {
      int64_t * attr_64 = (int64_t *) attr;
      attr_64[ipp] = value;
    } else {
        ERROR1("EnzoInitialIsolatedGalaxy.C // assign_attribute_value():",
        "Unknown particle attribute int type %d", type);
    }
  }
```

Seeing that this affects several other initializers and solvers (files listed below), **I wanted to get people's input on a more general solution that could be at the Cello level** before moving any further. Currently what I have in mind is a function that returns a templated type. My first attempt was the following
```c++
template <class T>
T * ParticleData::attribute_typed_array
(ParticleDescr * particle_descr, int it, int ia, int ib)
{
  T * array = NULL;
  int type = particle_descr->attribute_type(it,ia);
  if (type == type_float) {
    array = (float *) particle_descr->attribute_array(it,ia,ib);
  } else if (type == type_double) {
    array = (double *) particle_descr->attribute_array(it,ia,ib);
  }
  return array;
}
```
but I couldn't get it to work. There are many Stackoverflow posts ([example](https://stackoverflow.com/questions/17649136/function-which-is-able-to-return-different-types)) and other resources ([example](https://devblogs.microsoft.com/oldnewthing/20191106-00/?p=103066)) about this question, but I couldn't find a clear solution.

Affected files:
```
_data.hpp
enzo_EnzoComputeCicInterp.cpp
enzo_EnzoInitialFeedbackTest.cpp
enzo_EnzoInitialMergeStarsTest.cpp
enzo_EnzoInitialPm.cpp
enzo_EnzoMethodBackgroundAcceleration.cpp
enzo_EnzoMethodDistributedFeedback.cpp
enzo_EnzoMethodFeedback.cpp
enzo_EnzoMethodFire2Feedback.cpp
enzo_EnzoMethodMergeStars.cpp
enzo_EnzoMethodMergeStars.hpp
enzo_EnzoMethodPmDeposit.cpp
enzo_EnzoMethodPmUpdate.cpp
enzo_EnzoMethodStarMakerFIRE2.cpp
enzo_EnzoMethodStarMakerStochasticSF.cpp
enzo_EnzoRefineParticleMass.cpp
```

**Comment by stefanarridge** [comment]
I've encountered a similar problem before... my solution was to put `"default"` for the types of attributes in the input parameter file.

I've just had a look at the files in `input/IsolatedGalaxy`; it looks like they specify `"single"` or `"double"` as the types.

It would be good to have a more robust solution that doesn't rely on users having to know to put the right things in the parameter file. My preferred solution (which would solve some other problems as well), would be that particle attributes do not need to be specified in the input parameter file - only the particle types. The particle attributes, and the types of the attributes, would then be determined by the methods used.

**Comment by mabruzzo** [comment]
@jwise77 I think your general solution is a great idea! I've actually thought about suggesting this in the past.

In fact, as part of PR #151 I introduced something that looks a lot like your attempted solution for the `Field` class. The `Field::view` and `Field::coarse_view` methods both take the expected datatype of the field as a template parameter and returns the appropriate `CelloArray`. If the expected datatype doesn't match the actual datatype, the code aborts with an error.

At the time that I wrote the PR, I was thinking that a similar solution could be implemented for the other methods of `Field`, (like `Field::values`, `Field::unknowns`, and `Field::coarse_values`), but I decided not to pursue it at the time (the PR was already large enough). It should be really simple to do this - we could easily refactor some of the machinery in `FieldDescr` I implemented for `Field::view` and `Field::coarse_view`. I was actually thinking about bringing this up next week to try to get buy-in from everybody next week (since it would alter a bunch of existing code).

At that time, I was also thinking we could implement something similar for `Particles`, (exactly like you are suggesting). But it was a little less clear to me what that would look like (since I'm less familiar with the underlying machinery). 

The reason that your snippet isn't currently working is that you are assigning `array` values of types in different branches of the if-statement. That's not supported in C++11 (I'm pretty sure `constexpr` if-statements would support something like your version of the code in C++17). A working version of your snippet might look like the following (I haven't actually tested this, but it's very similar to the ``FieldData::make_view_`` implementation in `data_FieldData.cpp`):

```c++
#include <type_traits>

template<class T>
bool verify_type_ (int attribute_type) {
  switch (attribute_type){
  case type_float: return std::is_same<T, float>::value;
  case type_double: return std::is_same<T, double>::value;
  }
  return false;
}

template <class T>
T * ParticleData::attribute_typed_array
(ParticleDescr * particle_descr, int it, int ia, int ib)
{
  T * array = NULL;
  if (!verify_type_<T>(particle_descr->attribute_type(it,ia))) {
    // we might want to make the error message
    ERROR("ParticleData::attribute_typed_array",
         "The attribute has an unexpected type");
  }
  return (T *) particle_descr->attribute_array(it,ia,ib);
}
```

As an aside, it might also be cool to make a method that returns a `CelloArray` since we could directly encode the appropriate stride into the array (This would involve a minor tweak to the `CelloArray` class).

@stefanarridge I think you also have a good idea (I think they are both good things to do). Your solution sounds a lot like the automatic field detection we have. I imagine it would be easy to introduce something similar for Particles.

**Comment by stefanarridge** [comment]
> Your solution sounds a lot like the automatic field detection we have.

Can you help me understand how this works exactly? I notice that in the constructor of `EnzoMethodGravity` there are several calls to `cello::define_field`, but when I have tried to do this in `Method`s that I have implemented, I have found that I still need to provide the names of the fields in the input parameter file. I have discussed this with Will (can't seem to tag him) and he seems to have a similar issue. Is it something to do with whether the field's values need to be explicitly initialized?

**Comment by jwise77** [comment]
@stefanarridge I think changing the particle attribute precision in the affected tests to `default` is a good short-term solution. I can do that shortly. I agree that we should let the initializer methods automatically setup the appropriate particle attributes, given a set of parameters. One example where I'd like different particle precisions is higher precision (double/quad) for positions than the other attributes, e.g. creation_time or metallicity. We could have this automatically setup with a parameter controlling the position precision.

@mabruzzo Thank you very much for the code and insight from your work on the `Field` class! This makes sense of what you're doing here. I like your idea of converting the arrays in `ParticleData` to `CellArray`s. We can discuss more next week.

See you all in Maynooth!

**Comment by stefanarridge** [comment]
>One example where I'd like different particle precisions is higher precision (double/quad) for positions than the other attributes, e.g. creation_time or metallicity. We could have this automatically setup with a parameter controlling the position precision.

Yes that's a good point. In fact, positions / coordinates of blocks and cells are always doubles, so it would make sense for particle positions to be doubles as well.
